File: src/main/java/com/github/solairerove/algs4/leprosorium/arrays/SpiralTraverse.java
package com.github.solairerove.algs4.leprosorium.arrays;

import java.util.ArrayList;
import java.util.List;

public class SpiralTraverse {
    public static void main(String[] args) {
        int[][] a = new int[][]{
                {1, 2, 3, 4},
                {12, 13, 14, 5},
                {11, 16, 15, 6},
                {10, 9, 8, 7}
        };

        System.out.println(spiralTraverse(a));
    }

    private static List<Integer> spiralTraverse(int[][] a) {
        var res = new ArrayList<Integer>();
        int startRow = 0, endRow = a.length - 1;
        int startCol = 0, endCol = a[0].length - 1;

        while (startRow <= endRow && startCol <= endCol) {
            for (int col = startCol; col <= endCol; col++)
                res.add(a[startRow][col]);

            for (int row = startRow + 1; row <= endRow; row++)
                res.add(a[row][endCol]);

            for (int col = endCol - 1; col >= startCol; col--) {
                if (startRow == endRow) break;
                res.add(a[endRow][col]);
            }

            for (int row = endRow - 1; row > startRow; row--) {
                if (startCol == endCol) break;
                res.add(a[row][startCol]);
            }

            startRow++;
            endRow--;
            startCol++;
            endCol--;
        }

        return res;
    }
}

File: src/main/kotlin/com/github/solairerove/algs4/leprosorium/arrays/SpiralTraverse.kt
package com.github.solairerove.algs4.leprosorium.arrays

fun main() {
    val matrix = listOf(
        listOf(8, 2, -2, -4, 1),
        listOf(3, 5, 10, 6, 4),
        listOf(4, 10, -2, 1, 9),
        listOf(6, 9, -4, 0, -1)
    )

    print(spiralTraverse(matrix)) // [8, 2, -2, -4, 1, 4, 9, -1, 0, -4, 9, 6, 4, 3, 5, 10, 6, 1, -2, 10]
}

// O(n) time | O(n) space
private fun spiralTraverse(arr: List<List<Int>>): List<Int> {
    val res = mutableListOf<Int>()
    var lowRow = 0
    var highRow = arr.size - 1
    var lowCol = 0
    var highCol = arr[0].size - 1

    while (lowRow <= highRow && lowCol <= highCol) {
        for (col in lowCol..highCol) {
            res.add(arr[lowRow][col])
        }

        for (row in lowRow + 1..highRow) {
            res.add(arr[row][highCol])
        }

        for (col in highCol - 1 downTo lowCol) {
            if (lowRow == highRow) break
            res.add(arr[highRow][col])
        }

        for (row in highRow - 1 downTo lowRow + 1) {
            if (lowCol == highCol) break
            res.add(arr[row][lowCol])
        }

        lowRow++
        highRow--
        lowCol++
        highCol--
    }

    return res
}
